Add new module for miRTrace

chuan-wang opened the pull request:
Many thanks to contributing to MultiQC!

Please fill in the appropriate checklist below (delete whatever is not relevant). These are the most common things I request on pull requests (PRs).

## If this PR is for a new module
 - [X] There is example tool output for tools in the https://github.com/ewels/MultiQC_TestData repository
 - [X] Code is tested and works locally (including with `--lint` flag)
 - [X] `CHANGELOG.md` is updated
 - [X] `README.md` is updated
 - [X] `docs/README.md` is updated with link to below
 - [X] `docs/modulename.md` is created
 - [X] Everything that can be represented with a plot instead of a table is a plot
 - [X] Report sections have a description and help text (with `self.add_section`)
 - [X] There aren't any huge tables with > 6 columns (explain reasoning if so)
 - [X] Module does not do any significant computational work

ewels commented:
Hmm, tests are failing: https://travis-ci.org/ewels/MultiQC/jobs/411186222#L2581-L2594

ewels commented on multiqc/modules/mirtrace/mirtrace.py:
```diff
@@ -0,0 +1,381 @@
+        self.complexity_data      =   self.ignore_samples(self.complexity_data)
+
+        # Warning when no files are found
+        if len(self.summary_data) == 0:
```
You should only have one of these checks and raise if there are none of any of the reports. We want to have the module run if we find anything, right? For example:

```python
if max(len(self.summary_data), len(self.length_data), len(self.contamination_data), len(self.complexity_data)) == 0:
    raise UserWarning
```

You can then use these checks later when adding the report sections, eg:

```python
if len(self.length_data) > 0:
    self.add_section (
        name = 'Read Length Distribution',
        anchor = 'mirtrace_length',
        plot = self.mirtrace_length_plot()
    )
```

ewels commented on multiqc/modules/mirtrace/mirtrace.py:
```diff
@@ -0,0 +1,381 @@
+    # Parse a miRTrace mirtrace-stats-length.tsv file
+    def parse_length(self, f):
+        try:
+            with open(f['fn'], 'rb') as file:
```
If you use `self.find_log_files('mirtrace/length', filehandles=True):` then MultiQC will give you already opened file handles. Then you can simplify this code a little by removing the `with` statement:

```python
try:
    reader = csv.reader(f['f'], delimiter="\t")

ewels commented on multiqc/modules/mirtrace/mirtrace.py:
```diff
@@ -0,0 +1,381 @@
+                body = {}
+                for row in reader:
+                    body[row[0]]=row[1:len(row)]
+        except ValueError as e:
```
I don't really see why you need to use `try`/`except` to catch this `ValueError` if then all you do is re-`raise` it. I think you can get rid of these two lines and the code will work in exactly the same way.

ewels commented on multiqc/modules/mirtrace/mirtrace.py:
```diff
@@ -0,0 +1,381 @@
+                self.add_data_source(f, s_name)
+                self.length_data[s_name] = parsed_data
+        else:
+            log.debug('No valid data for read length distribution')
```
It's good to specify which file this log message relates to (mention `f['fn']`)

ewels commented on multiqc/modules/mirtrace/mirtrace.py:
```diff
@@ -0,0 +1,381 @@
+                for length in body:
+                    parsed_data[length] = int(body[length][idx])
+                if s_name in self.length_data:
+                    s_name = s_name + '_' + ''.join(random.choice(string.ascii_uppercase + string.ascii_lowercase + string.digits) for _ in range(6))
```
Why do you need to do this? When would the `s_name` be the same? Usually in these cases we overwrite the previous data and flag a debug log message:
https://github.com/ewels/MultiQC/blob/dd262ff4bbc5926b2ace4135f4a7ed36c021085f/multiqc/modules/salmon/salmon.py#L45-L46

chuan-wang replied:
Hi, I was thinking about the situation that multiple input fastq files were anlayzed for one sample. It will not happen based on our routine, but maybe someone else will run into such situation.

ewels commented on multiqc/modules/mirtrace/mirtrace.py:
```diff
@@ -0,0 +1,381 @@
+        keys['low_phred']                   =   { 'color': '#a50026', 'name': 'Reads with low PHRED score' }
+
+        # Construct a data structure for the plot - duplicate the samples for read 1 and read 2
+        data = {}
```
There's no need to create this data dict here as the structure and keys are the same as `self.summary_data`. MultiQC will ignore any data keys not specified in `keys`, so you can just call the function with the raw data directly here:

```python
return bargraph.plot(self.summary_data, keys, config)
```

ewels commented on multiqc/modules/mirtrace/mirtrace.py:
```diff
@@ -0,0 +1,381 @@
+            'cpswitch_counts_label': 'Number of Reads'
+        }
+
+        return bargraph.plot(data, keys, config)
```
As above, you can just use `self.summary_data` directly instead of creating `data`

ewels commented on multiqc/modules/mirtrace/mirtrace.py:
```diff
@@ -0,0 +1,381 @@
+        """ Generate the miRTrace Contamination Check"""
+
+        # A library of 24 colors. Should be enough for this plot
+        color_lib = ['rgb(166,206,227)', 'rgb(31,120,180)', 'rgb(178,223,138)', 'rgb(51,160,44)', 'rgb(251,154,153)', 'rgb(227,26,28)', 'rgb(253,191,111)', 'rgb(255,127,0)', 'rgb(202,178,214)', 'rgb(106,61,154)', 'rgb(255,255,153)', 'rgb(177,89,40)', 'rgb(141,211,199)', 'rgb(255,255,179)', 'rgb(190,186,218)', 'rgb(251,128,114)', 'rgb(128,177,211)', 'rgb(253,180,98)', 'rgb(179,222,105)', 'rgb(252, ...
```
Any reason not to use the default colours here?

ewels replied:
Ah ok, makes sense. The default bar plotting colours are from HighCharts actually, not colorbrewer. But the same thing applies - only 10 defaults before cycling: https://api.highcharts.com/highcharts/colors

Fine to keep it as there's a reason 👍

ewels commented on multiqc/modules/mirtrace/mirtrace.py:
```diff
@@ -0,0 +1,381 @@
+        keys = OrderedDict()
+        for clade in self.contamination_data[self.contamination_data.keys()[0]]:
+            keys[clade] = { 'color': color_lib[idx], 'name': clade }
+            idx = idx + 1
```
Shorthand for this is:

```python
idx += 1
```

You can also loop over a list and get the keys directly with `enumerate`:

```python
for idx, clade in enumerate(self.contamination_data[self.contamination_data.keys()[0]]):
    keys[clade] = { 'color': color_lib[idx], 'name': clade }
```

Though, if you do want to keep the custom colours, you should reset `idx` once it goes past 24 so that the colours recycle (otherwise this will break with > 24 categories if it ever happens)

ewels commented on multiqc/modules/mirtrace/mirtrace.py:
```diff
@@ -0,0 +1,381 @@
+            'cpswitch_counts_label': 'Number of detected miRNA'
+        }
+
+        return bargraph.plot(data, keys, config)
```
Can just use `self.contamination_data` directly instead of making `data` again I think.

ewels commented on multiqc/utils/config_defaults.yaml:
```diff
@@ -295,7 +295,9 @@ module_order:
-
+    - mirtrace:
+        module_tag:
+            - smRNA
```
Maybe also `miRNA`

chuan-wang commented:
Hi Phil,

I have fixed the codes based on your comments. It seems that the
csv.reader() function does not work with file handlers (I tested f, f['f']
and f['fn'] but none worked). Therefore I switched to splitlines() insteads.

Best,
Chuan

On Thu, Aug 2, 2018 at 2:02 PM, Phil Ewels <notifications@github.com> wrote:

> *@ewels* commented on this pull request.
> ------------------------------
>
> In multiqc/modules/mirtrace/mirtrace.py
> <https://github.com/ewels/MultiQC/pull/809#discussion_r207199940>:
>
> > +        config = {
> +            'id': 'mirtrace_rna_categories_plot',
> +            'title': 'miRTrace: RNA Categories',
> +            'ylab': '# Reads',
> +            'cpswitch_counts_label': 'Number of Reads'
> +        }
> +
> +        return bargraph.plot(data, keys, config)
> +
> +
> +    # miRTrace Contamination Check
> +    def mirtrace_contamination_check(self):
> +        """ Generate the miRTrace Contamination Check"""
> +
> +        # A library of 24 colors. Should be enough for this plot
> +        color_lib = ['rgb(166,206,227)', 'rgb(31,120,180)', 'rgb(178,223,138)', 'rgb(51,160,44)', 'rgb(251,154,153)', 'rgb(227,26,28)', 'rgb(253,191,111)', 'rgb(255,127,0)', 'rgb(202,178,214)', 'rgb(106,61,154)', 'rgb(255,255,153)', 'rgb(177,89,40)', 'rgb(141,211,199)', 'rgb(255,255,179)', 'rgb(190,186,218)', 'rgb(251,128,114)', 'rgb(128,177,211)', 'rgb(253,180,98)', 'rgb(179,222,105)', 'rgb(252,205,229)', 'rgb(217,217,217)', 'rgb(188,128,189)', 'rgb(204,235,197)', 'rgb(255,237,111)']
>
> Ah ok, makes sense. The default bar plotting colours are from HighCharts
> actually, not colorbrewer. But the same thing applies - only 10 defaults
> before cycling: https://api.highcharts.com/highcharts/colors
>
> Fine to keep it as there's a reason 👍
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ewels/MultiQC/pull/809#discussion_r207199940>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARdKNOM2mVOdatBRRgHXaTwKgCLuRhNYks5uMuo8gaJpZM4Vr4RU>
> .
>
